File: examples/batch_analysis.m
function batch_analysis(data_folder, output_folder, varargin)
% BATCH_ANALYSIS - 批量分析多个 ABF 文件
%
% 用法:
%   batch_analysis(data_folder)
%   batch_analysis(data_folder, output_folder)
%   batch_analysis(data_folder, output_folder, 'threshold', 3.5)
%
% 输入:
%   data_folder    - 包含 ABF 文件的文件夹路径
%   output_folder  - 输出文件夹（可选，默认为 data_folder/results）
%   varargin       - 传递给 mepsc_analyzer 的参数
%
% 示例:
%   batch_analysis('./data')
%   batch_analysis('./data', './results', 'threshold', 4.0)

% 设置输出文件夹
if nargin < 2 || isempty(output_folder)
    output_folder = fullfile(data_folder, 'results');
end

if ~exist(output_folder, 'dir')
    mkdir(output_folder);
end

% 找到所有 ABF 文件
files = dir(fullfile(data_folder, '*.abf'));

if isempty(files)
    error('在 %s 中未找到 ABF 文件', data_folder);
end

fprintf('找到 %d 个 ABF 文件\n', length(files));
fprintf('输出目录: %s\n\n', output_folder);

% 存储所有汇总结果
all_summaries = struct();
success_count = 0;

% 分析每个文件
for i = 1:length(files)
    abf_file = fullfile(files(i).folder, files(i).name);
    fprintf('[%d/%d] 处理: %s\n', i, length(files), files(i).name);
    
    try
        % 添加父目录到路径
        addpath(fileparts(fileparts(mfilename('fullpath'))));
        
        % 运行分析
        results = mepsc_analyzer(abf_file, varargin{:});
        
        % 保存结果
        success_count = success_count + 1;
        all_summaries(success_count) = results.summary;
        
        fprintf('  ✓ 检测到 %d 个事件\n', results.summary.num_events);
        fprintf('  ✓ 平均振幅: %.2f pA\n\n', results.summary.mean_amplitude_pA);
        
    catch ME
        fprintf('  ✗ 错误: %s\n\n', ME.message);
        continue;
    end
end

% 保存合并的汇总结果
if success_count > 0
    summary_table = struct2table(all_summaries);
    summary_path = fullfile(output_folder, 'all_files_summary.csv');
    writetable(summary_table, summary_path);
    
    fprintf('\n所有文件处理完成!\n');
    fprintf('合并的汇总结果已保存到: %s\n', summary_path);
    
    % 显示统计信息
    fprintf('\n=== 整体统计 ===\n');
    fprintf('成功分析的文件数: %d\n', success_count);
    fprintf('平均事件频率: %.2f Hz\n', mean([all_summaries.event_frequency_Hz]));
    fprintf('平均振幅: %.2f pA\n', mean([all_summaries.mean_amplitude_pA]));
else
    fprintf('\n警告: 没有成功分析任何文件\n');
end

end

File: examples/batch_analysis.py
#!/usr/bin/env python3
"""
批量分析多个 ABF 文件的示例脚本

用法:
    python batch_analysis.py /path/to/data/folder
"""

import os
import sys
import glob
import pandas as pd
from pathlib import Path

# 添加父目录到路径以导入 mepsc_analyzer
sys.path.insert(0, str(Path(__file__).parent.parent))
from mepsc_analyzer import mEPSCAnalyzer


def batch_analyze(data_folder, output_folder=None, **kwargs):
    """
    批量分析文件夹中的所有 ABF 文件
    
    参数:
        data_folder: 包含 ABF 文件的文件夹路径
        output_folder: 输出文件夹（默认为 data_folder/results）
        **kwargs: 传递给 detect_events 的参数
    """
    # 设置输出文件夹
    if output_folder is None:
        output_folder = os.path.join(data_folder, 'results')
    os.makedirs(output_folder, exist_ok=True)
    
    # 找到所有 ABF 文件
    abf_files = glob.glob(os.path.join(data_folder, '*.abf'))
    
    if len(abf_files) == 0:
        print(f"错误: 在 {data_folder} 中未找到 ABF 文件")
        return
    
    print(f"找到 {len(abf_files)} 个 ABF 文件")
    print(f"输出目录: {output_folder}\n")
    
    # 存储所有汇总结果
    all_summaries = []
    
    # 分析每个文件
    for i, abf_file in enumerate(abf_files, 1):
        print(f"[{i}/{len(abf_files)}] 处理: {os.path.basename(abf_file)}")
        
        try:
            # 创建分析器
            analyzer = mEPSCAnalyzer(abf_file)
            
            # 估计基线和噪声
            analyzer.estimate_baseline()
            analyzer.estimate_noise()
            
            # 检测事件
            analyzer.detect_events(**kwargs)
            
            # 分析所有事件
            analyzer.analyze_all_events()
            
            # 获取汇总统计
            summary = analyzer.get_summary_statistics()
            all_summaries.append(summary)
            
            # 保存结果
            analyzer.save_results(output_folder)
            
            # 生成并保存分析图
            base_name = os.path.splitext(os.path.basename(abf_file))[0]
            plot_path = os.path.join(output_folder, f"{base_name}_analysis.png")
            analyzer.plot_overview(save_path=plot_path)
            
            print(f"  ✓ 检测到 {len(analyzer.event_indices)} 个事件")
            print(f"  ✓ 平均振幅: {summary['平均振幅 (pA)']:.2f} pA\n")
            
        except Exception as e:
            print(f"  ✗ 错误: {str(e)}\n")
            continue
    
    # 合并所有汇总结果
    if all_summaries:
        combined_summary = pd.DataFrame(all_summaries)
        summary_path = os.path.join(output_folder, 'all_files_summary.csv')
        combined_summary.to_csv(summary_path, index=False)
        print(f"\n所有文件处理完成!")
        print(f"合并的汇总结果已保存到: {summary_path}")
        
        # 显示统计信息
        print("\n=== 整体统计 ===")
        print(f"成功分析的文件数: {len(all_summaries)}")
        print(f"平均事件频率: {combined_summary['事件频率 (Hz)'].mean():.2f} Hz")
        print(f"平均振幅: {combined_summary['平均振幅 (pA)'].mean():.2f} pA")
    else:
        print("\n警告: 没有成功分析任何文件")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("用法: python batch_analysis.py <data_folder> [output_folder]")
        print("\n示例:")
        print("  python batch_analysis.py ./data")
        print("  python batch_analysis.py ./data ./results")
        sys.exit(1)
    
    data_folder = sys.argv[1]
    output_folder = sys.argv[2] if len(sys.argv) > 2 else None
    
    # 运行批量分析
    batch_analyze(
        data_folder,
        output_folder,
        threshold_factor=3.5,
        min_interval_ms=5
    )
